### DISCUSSION TITLE
Opentelemetry support for cherrypy

### DISCUSSION THREAD

**Discussion by roshnidevadiga** [cherrypy/cherrypy#2009] [Q&A]
I have a cherrypy application and my org now wants to implement opentelemetry tracking. I am using the opentelemetry middleware for wsgi appilcation. But I am facing issues initializing it. Can the community help me how to fix this? Also if anyone has any pointers for this implementation, it will be really helpful

**Comment by webknjaz** [comment]
You're going to need to post more concrete examples of what you tried. Also, asking in the Matrix (Gitter) room might be a good idea.

**Reply by roshnidevadiga** [reply]
Here is how I am trying to instrument the cherrypy app
`from opentelemetry.instrumentation.wsgi import OpenTelemetryMiddleware
.
.
.
def request_filter(environ):
    # Add the URLs or paths you want to exclude
    excluded_paths = ['/abc', '/xyz', '/sadsd']
    path = environ.get('PATH_INFO', '/')
    return not any(path.startswith(excluded) for excluded in excluded_paths)

app = cherrypy.tree.mount(root, "/", config=config) #this is my app
app.wsgiapp.pipeline.append(('otel_middleware', OpenTelemetryMiddleware(app, request_hook=request_filter)))  #instrumenting the app. This is where it fails

`


The error that it throws is something like this
`traceback (most recent call last): File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cheroot/server.py", line 1279, in communicate req.respond() File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cheroot/server.py", line 1083, in respond self.server.gateway(self).respond() File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cheroot/wsgi.py", line 143, in respond response = self.req.server.wsgi_app(self.env, self.start_response) File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cherrypy/_cptree.py", line 302, in call return app(environ, start_response) File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cherrypy/_cptree.py", line 168, in call return self.wsgiapp(environ, start_response) File "/usr/local/pyenv/versions/3.9.1/lib/python3.9/site-packages/cherrypy/_cpwsgi.py", line 433, in call head = callable(head, **conf) TypeError: call() missing 1 required positional argument: 'start_response'`

**Reply by roshnidevadiga** [reply]
I also tried a custom tool implementation as follows in which I added a custom hook in the lifecycle to mimic the opentelemetry middleware. But the issue I am facing here is that the server span ends immediately in a few microseconds( maybe due to the way I am initializing the span

```
class OpenTelemetryTool(cherrypy.Tool):
    def __init__(self):
        self.excluded_urls = ['/asdsd','/sdsd',"/asdsd", "/sadsad"]
        cherrypy.Tool.__init__(self, 'before_handler', self.start_span, priority=10)

    def start_span(self):
        request_url = cherrypy.request.path_info
        # if any(re.match(url, request_url) for url in self.excluded_urls):
        if any(url in request_url for url in self.excluded_urls):
            # If the URL is excluded, return early without starting a span
            return
        self.tracer = trace.get_tracer(__name__)
        self.span = self.tracer.start_as_current_span(f"{cherrypy.request.method} {cherrypy.request.path_info}", kind=trace.SpanKind.SERVER)
```
       with trace.use_span(self.span, end_on_exit=False):
            # Store the current span in the request
            cherrypy.request.current_span = self.span
       

            # Inject the current span context into a dictionary and store it in the request
            headers = {}
            propagate.inject(headers)
            SyncLog.info("################ {}".format(headers.get("traceparent")))
            cherrypy.request.traceparent = headers.get("traceparent")

            cherrypy.request.hooks.attach('before_finalize', self.stop_span, priority=80)


     def stop_span(self):
        if self.span:
            self.span.set_attribute("http.status_code", cherrypy.response.status or 200)
            self.span.set_attribute("http.method", cherrypy.request.method)
            # if cherrypy.response.status >= 400:
            #     self.span.set_status(Status(StatusCode.ERROR, str(cherrypy.response.status)))
            self.span.end()
        else:
            return

cherrypy.tools.opentelemetry = OpenTelemetryTool()
```
In the config I have set "tools.opentelemetry.on": True,

**Reply by roshnidevadiga** [reply]
Any updates here? I am still stuck on this
